worker/upload.go

```diff
@@ -771,12 +811,24 @@ outer:
 			// execute it
 			var root types.Hash256
 			start := time.Now()
-			err := rl.withRevision(upload.ctx, defaultRevisionFetchTimeout, u.fcid, u.hk, u.siamuxAddr, lockingPriorityUpload, u.blockHeight(), func(rev types.FileContractRevision) error {
+			fcid, _, _ := u.contractInfo()
+			err := rl.withRevision(upload.ctx, defaultRevisionFetchTimeout, fcid, u.hk, u.siamuxAddr, lockingPriorityUpload, u.blockHeight(), func(rev types.FileContractRevision) error {
+				if rev.RevisionNumber == math.MaxUint64 {
```

Comment:
Why not handle this in `execute`? Before you call `UploadSector` you check the revision, if it is at the max you use the manager to refresh the uploader and then you call `UploadSector` like you usually would.

Ideally this also gets rid of the `requeue` method since you don't need to requeue. The request already made it through the queue so we might as well use it.

Sure you end up tracking the time it took to ask the bus for the contract but that's nanoseconds or milliseconds depending on your setup.

Reply:
I initially wanted to inline the renew but we have to fetch the revision again right. So `requeue` and handle as usual seemed like the cleanest solution. Did you miss that perhaps? Do you still think inlining it is cleaner if you have to fetch the revision after switching out the contract?